src/cdk/a11y/input-modality/input-modality-detector.ts

```diff
@@ -0,0 +1,187 @@
+  readonly inputModalityChange: Observable<InputModality>;
+
+  /** Returns the last detected / current input modality. */
+  get inputModality(): InputModality {
+    return this._inputModality;
+  }
+
+  /** The underlying ReplaySubject that emits input modality changes. */
+  private readonly _change = new ReplaySubject<InputModality>(1);
+
+  /** The current / last detected input modality by this service. */
+  private _inputModality: InputModality = null;
+
+  /** Options for this InputModalityDetector. */
+  private readonly _options: InputModalityDetectorOptions;
+
+  /**
+   * The timestamp of the last touch input modality. Used to determine whether mousedown events
+   * should be attributed to mouse or touch.
+   */
+  private _lastTouchMs = 0;
+
+  /**
+   * Handles keyboard events. Must be an arrow function in order to preserve the context when it
+   * gets bound.
+   */
+  private _onKeydown = (event: KeyboardEvent) => {
+    // If this is one of the keys we should ignore, then ignore it and don't update the input
+    // modality to keyboard.
+    if (this._options?.ignoreKeys?.some(key => key === event.key)) { return; }
+
+    this._updateInputModality('keyboard');
+  }
+
+  /**
+   * Handles mouse events. Must be an arrow function in order to preserve the context when it gets
+   * bound.
+   */
+  private _onMousedown = (event: MouseEvent) => {
+    if (isFakeMousedownFromScreenReader(event)) { return; }
+
+    // Touches trigger both touch and mouse events, so we need to distinguish between mouse events
+    // that were triggered via mouse vs touch. To do so, check if the mouse event occurs closely
+    // after the previous touch event.
+    if (Date.now() - this._lastTouchMs < TIME_AFTER_TOUCH_MS) { return; }
+
+    this._updateInputModality('mouse');
+  }
+
+  /**
+   * Handles touchstart events. Must be an arrow function in order to preserve the context when it
+   * gets bound.
+   */
+  private _onTouchstart = (event: TouchEvent) => {
+    if (isFakeTouchstartFromScreenReader(event)) { return; }
+
+    // Store the timestamp of this touch event, as it's used to distinguish between mouse events
+    // triggered via mouse vs touch.
+    this._lastTouchMs = Date.now();
+
+    this._updateInputModality('touch');
+  }
+
+  constructor(
+    platform: Platform,
+    @Inject(DOCUMENT) document: Document,
+    @Optional()
+    @Inject(INPUT_MODALITY_DETECTOR_OPTIONS)
+    options?: InputModalityDetectorOptions,
+  ) {
+    // If we're not in a browser, this service should do nothing, as there's no relevant input
+    // modality to detect.
+    if (!platform.isBrowser) { return; }
+
+    this._options = {
+      ...INPUT_MODALITY_DETECTOR_DEFAULT_OPTIONS,
+      ...options,
+    };
+
+    this.inputModalityChange =  this._change.pipe(distinctUntilChanged());
+
+    // Add the event listeners used to detect the user's input modality.
+    document.addEventListener('keydown', this._onKeydown, opts);
```

Comment:
All of these events should be bound outside of the Angular zone.

Reply:
Thanks, I made this change, but it's causing some problems in the dev-app demo and Stackblitz. If you take a look at the Stackblitz (I updated with the new implementation of binding these event listeners outside of the ngZone), you'll notice that the input modality no longer updates in the view (though of course the subscription still emits). Without knowing too much about zones my it seems like Angular isn't running CD because it doesn't think anything is changing. Is this expected behavior? If so, I'll need to update the dev-app demo accordingly.

Reply:
If the listener is bound outside of the zone, you'll have to use `NgZone.run` when updating the view, otherwise change detection won't run. It's a little cumbersome, but we may end up triggering unnecessary change detections otherwise.

Reply:
Gotcha, done.